compiler/packages/babel-plugin-react-compiler/src/Optimization/OutlineFunctions.ts

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { HIRFunction } from "../HIR";
+
+export function outlineFunctions(fn: HIRFunction): void {
+  for (const [, block] of fn.body.blocks) {
+    for (const instr of block.instructions) {
+      const { value } = instr;
+
+      if (
+        value.kind === "FunctionExpression" ||
+        value.kind === "ObjectMethod"
+      ) {
+        // Recurse in case there are inner functions which can be outlined
+        outlineFunctions(value.loweredFunc.func);
```

Comment:
If we choose to outline the parent function, is there any benefit in outlining its nested functions?

Reply:
Probably minimal, but I have to imagine it can still be slightly faster to avoid creating a closure unnecessarily.